### PR TITLE
WIP: Python current work directory no longer changes when starting and stopping aggregator and collaborator

### PR DESCRIPTION
**Pull request by aleksandr-mokrov**
Change working dir structure (client_id was added to path).
An opportunity to pass working directory path to aggregator and collaborator was added.
Python current work directory no longer changes when starting and stopping aggregator and collaborator form director and envoy.
Collaborator working directory creation was moved from director_client to envoy.

### CODE REVIEW

File: openfl/component/director/director.py [hunk 1]
```diff
@@ -64,20 +64,25 @@ async def set_new_experiment(self, *, experiment_name: str, sender_name: str,
...
         ))
 
         logger.info(f'New experiment {experiment_name} for '
                     f'collaborators {collaborator_names}')
         for col_name in collaborator_names:
             queue = self.col_exp_queues[col_name]
-            await queue.put(experiment_name)
+            await queue.put({
+                'experiment_name': experiment_name,
+                'client_id': sender_name
```
Comment (igor-davidyuk) [inline openfl/component/director/director.py:84]: Don't we disclose the experiment owner's identity to envoys here? Is it necessary?

File: openfl/component/director/director.py [hunk 2]
```diff
@@ -64,20 +64,25 @@ async def set_new_experiment(self, *, experiment_name: str, sender_name: str,
         """Set new experiment."""
         # TODO: save to file
         self.experiment_data[experiment_name] = data
+        experiment_work_dir = Path(f'{os.getcwd()}/{sender_name}/{experiment_name}')
```
Comment (igor-davidyuk) [inline openfl/component/director/director.py:67]: We discussed making timestamps-names for experiment working directories, is it a better idea?

File: openfl/component/envoy/envoy.py [hunk 3]
```diff
@@ -46,20 +54,21 @@ def run(self):
         """Run of the envoy working cycle."""
         while True:
             try:
-                # Workspace import should not be done by gRPC client!
-                experiment_name = self.director_client.get_experiment_data()
+                experiment_name, client_id = self.director_client.wait_experiment()
```
Comment (igor-davidyuk) [inline openfl/component/envoy/envoy.py:57]: experiment owner's id on envvoys side, is it really needed?

File: openfl/component/envoy/envoy.py [hunk 4]
```diff
@@ -46,20 +54,21 @@ def run(self):
...
+                experiment_work_dir = Path(f'{os.getcwd()}/{client_id}/{experiment_name}')
+                data_stream = self.director_client.get_experiment_data(experiment_name)
+                self.create_workspace(experiment_name, experiment_work_dir, data_stream)
```
Comment (igor-davidyuk) [inline openfl/component/envoy/envoy.py:60]: This is a strange signature for the `create_workspace` method. Why it accepts data_stream? Its only responsibility should be to create the workspace, not reading the stream.

File: openfl/component/envoy/envoy.py [hunk 5]
```diff
@@ -73,26 +82,74 @@ def send_health_check(self):
...
+            if experiment_work_dir in sys.path:
+                sys.path.remove(str(experiment_work_dir))
+
+    @staticmethod
+    def remove_workspace(experiment_name):
+        """Remove the workspace."""
+        shutil.rmtree(experiment_name, ignore_errors=True)
+
+    @staticmethod
+    def create_workspace(experiment_name, experiment_work_dir, response_iter):
```
Comment (igor-davidyuk) [inline openfl/component/envoy/envoy.py:118]: Again, we have this functionality in three places: Director, Envoy, and the old CLI's `workspace import`.
We may think of creating one helper function and use it everywhere.

File: openfl/component/envoy/envoy.py [hunk 6]
```diff
@@ -73,26 +82,74 @@ def send_health_check(self):
...
+                else:
+                    raise Exception('Broken archive')
+
+        shutil.unpack_archive(arch_name, experiment_work_dir)
+        os.remove(arch_name)
+
+        requirements_filename = f'./{experiment_work_dir}/requirements.txt'
+
+        if os.path.isfile(requirements_filename):
+            attempts = 3
```
Comment (igor-davidyuk) [inline openfl/component/envoy/envoy.py:140]: a hardcoded variable.

File: openfl/federated/plan/plan.py [hunk 7]
```diff
@@ -64,7 +64,8 @@ def ignore_aliases(self, data):
 
     @staticmethod
     def parse(plan_config_path: Path, cols_config_path: Path = None,
-              data_config_path: Path = None, resolve=True):
+              data_config_path: Path = None, resolve=True,
+              working_dir: Path = None):
```
Comment (igor-davidyuk) [inline openfl/federated/plan/plan.py:68]: It feels like we are polluting the Plan's code with folder management.
We removed two `os.chdir()` lines and then wrote like 150 more.

**Review by igor-davidyuk — CHANGES_REQUESTED** [review]

File: openfl/component/director/director.py (hunk 2 above)
Comment (dmitryagapov) [reply openfl/component/director/director.py:67]: Maybe we can create constant path to save experiments data, e.g. `~/.openfl/experiments/`?

**Pull request closed without merging**